add support for proxied users to proxy-pki strategy

jrassa opened the pull request:
Supports proxied users via `x-proxied-user-dn` HTTP header.

Currently implementation logs in the proxied user, but all subsequent actions will be performed by the primary user.  However, primary users roles/accesses will be the intersection of the 2 users.

Not sure if this is the best/proper implmentation, but at minimum should serve as a good starting point for extension.

jrassa committed:
add support for proxied users to proxy-pki strategy

brianghig reviewed (changes requested):
@jrassa from a code inspection, this looks functionally correct.

However, if proxying, the "primary user" is the server, right? If that's the case, I'd recommend we flip the last part in order to send the proxied user's credentials through to the rest of the system. In this case, an actual user navigating via another system will still see their data (albeit, potentially less of it due to the authorization intersection).

In this case, the proxied user would see things like their own favorites or saved searches. Otherwise, a bunch of people using the same "primary user" proxy could be using the same preferences, saved searches, etc.

What do you think of that approach?

jrassa commented:
@brianghig 
> I'd recommend we flip the last part in order to send the proxied user's credentials through to the rest of the system

I agree, but we need to first address the following

> However, if proxying, the "primary user" is the server, right?

To be clear, there are two different layers of proxying that are occurring.  Regardless of the changes in this PR, when running with 'proxy-pki', there needs to be a server (we typically use nginx
) that is receiving the incoming request, processing the pki certificate, and then proxying the request 
to node-rest-starter.  This internal proxying establishes the "primary user".

With this change, we are now adding a way to allow the "primary user" to proxy a request on behalf of a different user.  However, there is no guarantee that the "primary user" is a server.  As is, any user with enough knowledge, could use the dev tools in the browser to add the appropriate header and proxy their requests as a different user.

So the question is, how do we want to lock down this new proxy ability?

brianghig commented:
@jrassa very good question. That seems like an "allow proxy" role per user would help us to secure against that behavior.

The use case that I'm working is:

User
\> External System (with certificate)
\> \> NginX (ours with certificate)
\> \> \> Node Rest Server (ours)

In this case:

* the "External System" would be the trusted system (verified via the "allow proxy" role), and
* the "User" would be the "primary user" who the rest of the queries would run as.

Does that make sense?

jrassa committed:
proxied users: add check on primary user to see if they are authorized to proxy

jrassa commented:
@brianghig Makes sense.  I pushed an update.  Added a 'canProxy' boolean flag to the user and added a check for the flag and flipped it at the end so that the proxied user is the active one.

brianghig committed:
Merge branch 'develop' into proxy-user-support

brianghig commented on src/lib/strategies/proxy-pki.js:
```diff
@@ -125,112 +126,133 @@ function createUser(dn, acUser) {
-			return done(err);
-		});
-
+			if (proxiedUser === null) {
```
@jrassa thinking about some more use cases here... should we fail / throw an error if the request contains a proxied user DN, but the primary requestor is not allowed to proxy? That may be a more appropriate response for a server that is misconfigured, potentially avoiding any confusion about their ability to proxy.

brianghig replied:
On a side note, it looks like our unit tests for the node-rest-starter are ignoring errors from missing email templates in several places. It would be especially useful to have some test coverage around this piece of functionality (since it deals with authentication and authorization), so I'm going to work up a new branch to get those back online.

brianghig committed:
Updating package.json to resolve security vulnerabilities

brianghig committed:
Fixing user-authentication service references to email templates for welcome email and signup alert

brianghig committed:
Merge branch 'feature/tests-and-patches' into proxy-user-support

brianghig committed:
Rejecting requests to proxy for users who are not allowed to proxy

brianghig committed:
Cleaning up lint errors for hasOwnProperty

brianghig committed:
Merge branch 'feature/tests-and-patches' into proxy-user-support
